dxr/plugins/rust/refs.py

```diff
@@ -0,0 +1,339 @@
+from warnings import warn
+
+from dxr.lines import Ref
+# import ALL THE THINGS
+from dxr.plugins.rust.menu import (
+    jump_to_target_from_decl, jump_to_definition_menu_item,
+    jump_to_trait_method_menu_item, generic_function_menu, generic_variable_menu, truncate_value,
+    generic_type_menu, jump_to_module_definition_menu_item, generic_module_menu,
+    jump_to_alias_definition_menu_item, jump_to_crate_menu_item, find_references_menu_item,
+    std_lib_links_menu, jump_to_module_declaration_menu_item, jump_to_type_declaration_menu_item,
+    jump_to_variable_declaration_menu_item, jump_to_function_declaration_menu_item,
+    trait_impl_menu_item)
+
+
+def trim_dict(dictionary, keys):
+    """Return a new dict with given keys set from dictionary arg, or None if dictionary is falsy.
+
+    A key listed in keys but not in the dictionary will not be set in the returned dict.
+    """
+
+    if dictionary:
+        return dict((key, dictionary[key]) for key in keys if key in dictionary)
+
+
+class _RustRef(Ref):
+    plugin = 'rust'
+
+    def __init__(self, tree_config, menu_data, tree_index=None, hover=None, qualname=None, qualname_hash=None):
+        super(_RustRef, self).__init__(tree_config, None, hover, qualname, qualname_hash)
+        if tree_index:
+            # If we are at index time, then prepare menu_data for ES insertion.
+            # Otherwise we know we already have menu_data populated from previous preparation.
+            self.menu_data = self.prepare_menu_data(tree_index, menu_data)
+        else:
+            self.menu_data = menu_data
+
+    def prepare_menu_data(self, tree_index, datum):
+        """Return necessary data for insertion to ES and later retrieval via menu_items.
+
+        This function should handle any logic that involves tree_index, as it's
+        an index-time construct. Data can be any shape (dicts, lists, etc.) as long as
+        menu_items knows what it's dealing with.
+        """
+
+        raise NotImplementedError
+
+
+class _KeysFromDatum(_RustRef):
+    """A Ref superclass that only holds onto certain keys of a datum dict.
+
+    :classvar keys: a list of dict keys which will be kept in the data to pass to ES.
+    """
+
+    def __init__(self, tree_config, datum, tree_index=None, hover=None, qualname=None, qualname_hash=None):
+        super(_KeysFromDatum, self).__init__(tree_config, datum, tree_index, hover, qualname, qualname_hash)
+
+    def prepare_menu_data(self, tree_index, datum):
+        return trim_dict(datum, self.keys)
+
+
+class FunctionRef(_RustRef):
+    def prepare_menu_data(self, tree_index, datum):
+        if 'declid' in datum and datum['declid'] in tree_index.data.functions:
+            # It's an implementation; find the decl:
+            decl = tree_index.data.functions[datum['declid']]
+            return [trim_dict(datum, ['qualname']), trim_dict(decl, ['file_name', 'file_line']), 0]
```

Comment:
Good point, it's an easy adjustment to only save qualname. generic_function_menu should really only take qualname as an argument, rather than a datum dict where it only uses qualname.